### PR TITLE
Introduce `ContextPathServicesBuilder` (context path part 3)

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

As a last part of refactoring, I propose that we introduce a separate `ContextPathServicesBuilder` which contains the actual logic to add a `ServiceConfig` to a `VirtualHostBuilder`. 

Previousy, the logic for updating was as follows:

```
ServerBuilder#service -> virtualHostTemplate#service -> virtualHostBuilder#addServiceConfigSetters
``` 

I propose that we introduce a `ContextPathServicesBuilder` as an additional layer when adding `ServiceConfigSetters`.

```
ServerBuilder#service -> virtualHostTemplate#service -> ContextPathServicesBuilder#service -> virtualHostBuilder#addServiceConfigSetters
``` 

The actual logic for creating `ServiceConfig`s will be moved to `ContextPathServicesBuilder`.
By doing so, I believe we can easily add an extension to add arbitrary paths like the following in the future:

i.e.
```java
ContextPathServicesBuilder<ServerBuilder> contextPath(String... contextPaths) {
  return new ContextPathServicesBuilder<>(this, defaultVirtualHostBuilder, contextPaths);
}
```

While in the process, I realized that `AnnotatedServiceConfigSetters annotatedService()` is used as the main way for creating annotated service bindings. For this reason, I've also created a `ContextPathAnnotatedServiceConfigSetters` and included the class in this patch.

final version: https://github.com/line/armeria/pull/5037/files

Modifications:

- Introduced a `ContextPathServicesBuilder` which contains logic for creating `ServiceConfigSetters`
- Migrated logic for creating `ServiceConfigSetters` from `VirtualHostBuilder` to `ContextPathServicesBuilder`
- Introduced a `ContextPathAnnotatedServiceConfigSetters` which overrides `AbstractAnnotatedServiceConfigSetters`. This is done so that the actual logic in `ContextPathServicesBuilder` can use the `annotatedService()` API

Result:

- Refactoring is done in preparation for introducing the `ContextPath` API.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+
+    @Override
+    public AbstractServiceBindingBuilder route() {
+        throw new UnsupportedOperationException("Not supported yet");
+    }
+
+    @Override
+    public AbstractBindingBuilder routeDecorator() {
+        throw new UnsupportedOperationException("Not supported yet");
+    }
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java:57]: Note that these are not called from anywhere yet. These will be implemented when this class is exposed publicly in the next stage

File: core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+ * @param <T> the original type which will be returned once the {@link ServiceConfig}
+ *            is built using {@link #and()}.
+ */
+final class ContextPathServicesBuilder<T> implements ServiceConfigsBuilder {
+
+    private final T parent;
+    private final VirtualHostBuilder virtualHostBuilder;
+
+    ContextPathServicesBuilder(T parent, VirtualHostBuilder virtualHostBuilder) {
+        this.parent = parent;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java:45]: Q: Is the generic `T` type used to take `VirtualHostBuilder` and `ServerBuilder`? If possible, I prefer defining the upper bound.

**Review by ikhoon — APPROVED** [review]
Thanks, @jrhee17!

File: core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java (hunk 2 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java:46]: Added a bound on `T extends ServiceConfigsBuilder`, let me know if this isn't what you meant 😄

File: core/src/main/java/com/linecorp/armeria/server/ContextPathAnnotatedServiceConfigSetters.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+    /**
+     * Registers the given service to {@link T} and returns the parent object.
+     *
+     * @param service annotated service object to handle incoming requests matching path prefix, which
+     *                can be configured through {@link AnnotatedServiceBindingBuilder#pathPrefix(String)}.
+     *                If path prefix is not set then this service is registered to handle requests matching
+     *                {@code /}
+     */
+    public ContextPathServicesBuilder<T> build(Object service) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/ContextPathAnnotatedServiceConfigSetters.java:58]: nit:
```suggestion
    ContextPathServicesBuilder<T> build(Object service) {
```

**Review by minwoox — APPROVED** [review]
Thanks!

**Pull request merged**